includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php

```diff
@@ -77,6 +77,112 @@ private function is_changing_payment_method_for_subscription() {
 		return false;
 	}
 
+	/**
+	 * Overrides process_payment() to handle payment change for subscription order.
+	 *
+	 * @param int $order_id Order ID to process the payment for.
+	 *
+	 * @return array|null An array with result of payment and redirect URL, or nothing.
+	 */
+	public function process_payment( $order_id ) {
+		if ( ! $this->is_changing_payment_method_for_subscription() ) {
+			return parent::process_payment( $order_id );
+		}
+
+		$order               = wc_get_order( $order_id );
+		$payment_information = $this->prepare_payment_information( $order );
+		$user                = $order->get_user() ?? wp_get_current_user();
+		$name                = sanitize_text_field( $order->get_billing_first_name() ) . ' ' . sanitize_text_field( $order->get_billing_last_name() );
+		$email               = sanitize_email( $order->get_billing_email() );
+		$customer_id         = $this->upsert_customer( $user, $name, $email );
+
+		try {
+			$intent = $this->payments_api_client->create_and_confirm_setup_intent(
+				$payment_information->get_payment_method(),
+				$customer_id
```

Comment:
Let's use `$this->upsert_customer( $user, $name, $email )` here  for better context scoping.

Reply:
Good idea. Will do.